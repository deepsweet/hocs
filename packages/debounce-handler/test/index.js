import React from 'react'
import { mount } from 'enzyme'

const Target = () => null

describe('debounceHandler', () => {
  let mockJustDebounce = null
  let debounceHandler = null

  beforeEach(() => {
    mockJustDebounce = jest.fn(cb => cb)

    jest.mock('just-debounce-it', () => mockJustDebounce)
    jest.resetModules()

    debounceHandler = require('../src/').default
  })

  afterAll(() => {
    jest.unmock('just-debounce-it')
  })

  it('should pass handler arguments through', () => {
    const EnhancedTarget = debounceHandler('testHandler')(Target)
    const mockTestHandler = jest.fn()
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler} />
    )
    const testHandler = wrapper.find(Target).prop('testHandler')

    testHandler()

    expect(mockTestHandler.mock.calls).toMatchSnapshot()
  })

  it('should call actual handler during lifecycle', () => {
    const EnhancedTarget = debounceHandler('testHandler')(Target)
    const mockTestHandlerInitial = jest.fn()
    const mockTestHandlerUpdated = jest.fn()
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandlerInitial} />
    )
    wrapper.setProps({testHandler: mockTestHandlerUpdated})
    const testHandler = wrapper.find(Target).prop('testHandler')

    testHandler()

    expect(mockTestHandlerInitial.mock.calls).toMatchSnapshot()
    expect(mockTestHandlerUpdated.mock.calls).toMatchSnapshot()
  })

  it('should call `e.persist()` if it has been passed', () => {
    const EnhancedTarget = debounceHandler('testHandler')(Target)
    const mockTestHandler = jest.fn()
    const mockPersist = jest.fn()
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler} />
    )
    const testHandler = wrapper.find(Target).prop('testHandler')

    testHandler({ persist: mockPersist })

    expect(mockTestHandler.mock.calls).toMatchSnapshot()
    expect(mockPersist.mock.calls).toMatchSnapshot()
  })

  it('should pass `delay` option to `just-debounce-it`', () => {
    const EnhancedTarget = debounceHandler('testHandler', 75)(Target)
    const mockTestHandler = jest.fn()
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler} />
    )
    const testHandler = wrapper.find(Target).prop('testHandler')

    testHandler()

    expect(mockJustDebounce.mock.calls).toMatchSnapshot()
  })

  it('should pass `leadingCall` option to `just-debounce-it`', () => {
    const EnhancedTarget = debounceHandler('testHandler', 75, true)(Target)
    const mockTestHandler = jest.fn()
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler} />
    )
    const testHandler = wrapper.find(Target).prop('testHandler')

    testHandler()

    expect(mockJustDebounce.mock.calls).toMatchSnapshot()
  })

  describe('display name', () => {
    const origNodeEnv = process.env.NODE_ENV

    afterEach(() => {
      process.env.NODE_ENV = origNodeEnv
    })

    it('should wrap display name in non-production env', () => {
      process.env.NODE_ENV = 'test'

      const EnhancedTarget = debounceHandler('test')(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper).toMatchSnapshot()
    })

    it('should not wrap display name in production env', () => {
      process.env.NODE_ENV = 'production'

      const EnhancedTarget = debounceHandler('test')(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper).toMatchSnapshot()
    })
  })
})
