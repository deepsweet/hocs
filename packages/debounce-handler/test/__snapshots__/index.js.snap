// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debounceHandler display name should not wrap display name in production env 1`] = `
<DebounceHandler>
  <Target
    test={[Function]}
  />
</DebounceHandler>
`;

exports[`debounceHandler display name should wrap display name in non-production env 1`] = `
<debounceHandler(Target)>
  <Target
    test={[Function]}
  />
</debounceHandler(Target)>
`;

exports[`debounceHandler should call \`e.persist()\` if it has been passed 1`] = `
Array [
  Array [
    Object {
      "persist": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
  ],
]
`;

exports[`debounceHandler should call \`e.persist()\` if it has been passed 2`] = `
Array [
  Array [],
]
`;

exports[`debounceHandler should call actual handler during lifecycle 1`] = `Array []`;

exports[`debounceHandler should call actual handler during lifecycle 2`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`debounceHandler should pass \`delay\` option to \`just-debounce-it\` 1`] = `
Array [
  Array [
    [Function],
    75,
    undefined,
  ],
]
`;

exports[`debounceHandler should pass \`leadingCall\` option to \`just-debounce-it\` 1`] = `
Array [
  Array [
    [Function],
    75,
    true,
  ],
]
`;

exports[`debounceHandler should pass handler arguments through 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;
